File: day_1/py/main.py
# Zion Amsalem Advent Of Code 2025 - Day 1

PATH = "../input.txt"
LOCK = 50

def right(lock: int, steps: int) -> int:
    # move the lock n steps right
    lock = lock + steps
    if lock >= 100:
        lock = lock - 100
    return lock

def left(lock: int, steps: int) -> int:
    # move the lock n steps left
    lock = lock - steps
    if lock < 0:
        lock = 100 + lock
    return lock

def tests() -> None:
    # tests is all you need
    lock = LOCK
    assert left(lock, 0) == 50
    assert left(lock, 49) == 1
    assert left(lock, 100) == 50
    assert left(lock, 51) == 99
    assert right(lock, 50) == 0
    assert right(lock, 51) == 1
    assert right(lock, 52) == 2
    assert right(lock, 100) == 50
    assert right(lock, 1) == 51
    assert right(lock, (549 % 100)) == 99
    assert tour_test(lock) == 2

def tour_test(lock: int) -> int:
    # simplest e2e test
    result = 0
    _actions = ["R10", "L60", "R101", "L1"]
    for action in _actions:
        steps = int(action[1:]) % 100
        if action[0] == "R":
            lock = right(lock, steps)
        elif action[0] == "L":
            lock = left(lock, steps)
        if lock == 0:
            result += 1
    return result

def load_file(path: str) -> list[str]:
    # load the file
    with open(path, "r", encoding="utf8") as file:
        return file.readlines()

def pick(file: list[str], lock: int) -> int:
    # pick the lock
    result = 0
    for action in file:
        steps = int(action[1:]) % 100
        if action[0] == "R":
            lock = right(lock, steps)
        elif action[0] == "L":
            lock = left(lock, steps)
        if lock == 0:
            result += 1
    return result

if __name__ == "__main__":
    tests()
    actions_list = load_file(PATH)
    secret = pick(actions_list,LOCK)
    print(f"the code for the lock is: {secret}")

File: day_1/c/main.c
// Zion Amsalem ADV 2025 Day 1 C version
#include <stdio.h>
#include <stdlib.h>

void move_right(int *lock, int steps){
    *lock = (*lock + (steps % 100));
    if (*lock >= 100){
        *lock = (*lock - 100);
    };
}

void move_left(int *lock, int steps){
    *lock = (*lock - (steps % 100));
    if (*lock < 0){
        *lock = (*lock + 100);
    };
}

FILE* load_file(const char *path){
    FILE *input = fopen(path, "r");
    if (!input) {
        perror("an error occurred while trying to open this file");
        return NULL;
    };
    return input;
}

int pick(int *lock, FILE* file){
    int result = 0;
    char line[256];

    while (fgets(line, sizeof(line), file)) {
        char direction = line[0];
        int steps = atoi(&line[1]);

        if (direction == 'R'){
            move_right(lock, steps);
        }else if (direction == 'L'){
            move_left(lock, steps);
        };
        if (*lock == 0){
            result++;
        }
    };
    fclose(file);
    return result;
}

int main(){
    int lock = 50;
    const char *path = "../input.txt";
    FILE* loaded_file = load_file(path);

    if (!loaded_file) {
        return 1;
    }
    int res = pick(&lock, loaded_file);
    printf("the result of the secret is: %d\n",res);
    return 0;
}


